lib/puppet/util/tagging.rb

```diff
@@ -40,7 +40,7 @@ def tag(*ary)
   # since that results in testing the same string twice
   #
   def tag_if_valid(name)
-    if name.is_a?(String) && name =~ ValidTagRegex
+    if name.is_a?(String) && name =~ ValidTagRegex && Puppet::Resource::MAIN != name
```

Comment:
We can remove that test and the one and only 'main' class that indeed is created (by the compiler), will then get tagged with 'main'. That in turn causes a unit test failure.

If you think that tagging the main class with main is OK, then I'll remove this special case and fix that unit test.